promdump: Automatic batch duration backoff

ionthegeek opened the pull request:
This PR implements automatic batch duration backoff for promdump. If Prometheus finds that a query would return too many samples, it will throw an error that can be caught. With these changes in place, the error will be caught and the batch duration will be halved (down to a minimum of 1s) and the dump retried.

ionthegeek committed:
Update to golang 1.17 and prometheus/client_golang 1.14

ionthegeek committed:
promdump: Implement backoff for batch size

This diff adds a backoff mechanism for batch size.

If Prometheus receives a query that would produce more than a certain number of samples, it will fail the query to protect itself from running out of memory and crashing. The Prometheus API returns HTTP 422 in these cases.

Unfortunately, the Prometheus golang client swallows the HTTP response code, which leaves us with limited options for detecting this condition. The code in this diff uses a RegEx match to scrape the error message but that's probably fine because this is the kind of utility where we can iterate rapidly if a change in the message breaks the scraping.

If the "too many samples" condition occurs, the batch duration will be halved, the existing data will be discarded, and the dump will be restarted from the beginning. If the batch size ever falls below the minimum of 1 second, the utility will throw a fatal error and exit.

eugeneckim commented on promdump/promdump.go:
```diff
@@ -43,6 +47,20 @@ func writeFile(values *[]*model.SampleStream, fileNum uint) error {
+	}
+	for _, file := range files {
+		err = os.Remove(file)
+		if err != nil {
```
Perhaps we can print something about the removes.  Per our conversation, maybe not ALL removes, but a summary of how many, etc.

ionthegeek replied:
Updated to address this feedback.

eugeneckim reviewed (approved):
Go go go (intended)
Notes from conversation:
- Consider printing information about the cleanup function for visibility.
- We had discussions about maximum number of files and backstopping it at potentially 5 digits.

ionthegeek committed:
promdump: Implement backoff for batch size

This diff adds a backoff mechanism for batch size.

If Prometheus receives a query that would produce more than a certain number of samples, it will fail the query to protect itself from running out of memory and crashing. The Prometheus API returns HTTP 422 in these cases.

Unfortunately, the Prometheus golang client swallows the HTTP response code, which leaves us with limited options for detecting this condition. The code in this diff uses a RegEx match to scrape the error message but that's probably fine because this is the kind of utility where we can iterate rapidly if a change in the message breaks the scraping.

If the "too many samples" condition occurs, the batch duration will be halved, the existing data will be discarded, and the dump will be restarted from the beginning. If the batch size ever falls below the minimum of 1 second, the utility will throw a fatal error and exit. If the batch count would exceed 99999, the utility will exit.

The utility now also logs result counts, file writes, and file removals.

na6vj reviewed (commented):
Any chance you could get rid of the outer "for allBatchesFetched == false {"  loop ?
The inner loop checks the condition "batch <= batches" anyway, making "allBatchesFetched" superfluous.

eugeneckim reviewed (approved):
Let's go!

ionthegeek commented:
> Any chance you could get rid of the outer "for allBatchesFetched == false {" loop ? The inner loop checks the condition "batch <= batches" anyway, making "allBatchesFetched" superfluous.

Unfortunately this isn't possible because we have to reset both the current batch count and total batch count if the batch duration changes and these are calculated up front before the inner loop starts. The only straightforward way to redo the whole shebang is to wrap it in an outer loop with an exit condition (`allBatchesFetched`) and break out of the inner loop if we have to recalculate.
